### DISCUSSION TITLE
Help on Step 1 Activity 2 | continuous-delivery-azure

### DISCUSSION THREAD

**Discussion by KittyChiu** [skills/.github#14] [Q&A]
**"On staging workflow update"** job is failing.  I have checked the permissions for 
* GITHUB_TOKEN - Allow GitHub Actions to create and approve pull requests
* GITHUB_TOKEN - Read and write permissions

![image](https://user-images.githubusercontent.com/42864823/181693748-ceb0938c-e1e2-429d-a5a0-16bae1a9ba41.png)

I think the automated tutorial script is missing adding `spinup-destroy.yml` into `azure-configuration` branch.  Before line 80 in `2-setup-azure-environment.yml`, you are missing some code there.

**Comment by slgraff** [comment]
I've experienced the same problem.

**Comment by AnttiHal** [comment]
Same problem here.

**Comment by markpatterson27** [comment]
Step 1 is titled 'Trigger a job based on labels' and step 2 starts with 'Nice work triggering a job on specific labels'. At no point in step 1 do you actually trigger a job based on a label, or am I missing something? The only thing you do in Step 1 is create an invalid workflow (because it has no steps), and create a plain PR for the new file. The intended learning outcome for step 1 isn't achieved (in step 1).

Am I missing something? Maybe activity 2 should include a valid workflow (a 'hello-world' step would do) and then include an activity 3 that triggers it.

**Comment by markpatterson27** [comment]
The error is because you can't create a PR with no commits in it. There must be a difference in history between HEAD and base refs.

It looks like there used to be example workflows in the course repo, one of which was copied into the workflows dir and commited by the `create-workflow-pr.sh` script. The demo scripts were taken out and the `create-workflow-pr.sh` script was [altered, taking out the 'Added' workflow file commit](https://github.com/skills/continuous-delivery-azure/commit/8e1a9bb8c3f81a5f4c19e034456a78b7af624ae6).

As @KittyChiu said, `spinup-destroy.yml` needs to be added back in somewhere and commited to the `azure-configuration` branch before the new PR is opened.

Should the creation of the file and PR be automated, or would it be better to add it as a user activity?

(The needed `spinup-destroy.yml` file can be found in the [course repo history](https://github.com/skills/continuous-delivery-azure/blob/e7fe690eee690c14bd7e1d54cd67a82fcc0752b6/.github/example-workflows/spinup-destroy.yml) if anyone is looking for it.)

**Comment by markpatterson27** [comment]
I've created a [PR](https://github.com/skills/continuous-delivery-azure/pull/20) for this issue [WIP].

At a guess, I assume the missing workflow was removed because adding a workflow file is out of scope for the default `GITHUB_TOKEN`. This leaves the options of either having the learner create a `workflow` scoped PAT and adding it to their course repo, or having the learner create the missing workflow themselves. If anyone has views on which learner activity would be better, I'd appreciate the input. Thank you.

**Reply by jhoffman3** [reply]
As a rank beginner with github actions and Azure, even with the help of your comments and PR, it wasn't possible for me to create and use a scoped PAT to get around the workflow perm errors for azure-configuration or staging-test branch issues. I ended up commenting them and performing those steps manually. Learned a lot from the experience but not enough, yet :)

**Comment by sgould420** [comment]
This exercise should be removed from the learning path at [https://learn.microsoft.com/en-us/training/modules/github-actions-cd/3-create-workflow-deploy-azure](url)
until it is fixed.

**Comment by nicholasdagostino** [comment]
Hello,

I encountered an error on step 4,5,6 on this tutorial as well. When attempting to deploy to staging and production the npm install and build webpack step failed with this error:
![image](https://user-images.githubusercontent.com/97560278/226082162-6b49f073-7a87-4c4f-86c0-e56701f827d3.png)

**Reply by markpatterson27** [reply]
It's a [known issue](https://github.com/skills/continuous-delivery-azure/issues/39). Quick fix is to pin the node version on the runner to node v16. Add the following to the `deploy-staging.yml` and `deploy-prod.yml` workflows, just before the `- name: npm install and build webpack` step:

```yml
          - uses: actions/setup-node@v3
            with:
              node-version: 16
```

i.e.: https://github.com/skills/continuous-delivery-azure/pull/40/commits/d5ac4dbf54d8eaae539f5b7b04c0e9353c622ecd in PR: https://github.com/skills/continuous-delivery-azure/pull/40

> **Note**: the tutorial was updated last month. The issues commented on at the start of this discussion thread are no longer relevant.

**Reply by nicholasdagostino** [reply]
yes it worked, this is what i added:

steps:
      - uses: actions/checkout@v3

      - name: Setup Node.js environment
        uses: actions/setup-node@v2
        with:
          node-version: '16.x'

      - name: npm install and build webpack
        run: |
          npm install
          npm run build

      - uses: actions/upload-artifact@v3
        with:
          name: webpack artifacts
          path: public/
